nemo/core/neural_graph/neural_graph.py

```diff
@@ -0,0 +1,882 @@
+# -*- coding: utf-8 -*-
+
+# =============================================================================
+# Copyright (c) 2020 NVIDIA. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+
+__all__ = [
+    'NeuralGraph',
+]
+
+from collections import OrderedDict, namedtuple
+from os import path
+from typing import Dict, Optional
+
+from ruamel.yaml import YAML
+
+from nemo.core import OperationMode
+from nemo.core.neural_graph.graph_inputs import GraphInputs
+from nemo.core.neural_graph.graph_outputs import GraphOutputs
+from nemo.core.neural_interface import NeuralInterface
+from nemo.core.neural_modules import NeuralModule
+from nemo.core.neural_types import NeuralPortNameMismatchError, NeuralType
+from nemo.package_info import __version__ as nemo_version
+from nemo.utils import logging
+from nemo.utils.connection import Connection, StepModulePort
+
+YAML = YAML(typ='safe')
+
+
+class NeuralGraph(NeuralInterface):
+    """
+        Neural Graph class stores dynamically defined graphs of connected Neural Modules.
+    """
+
+    def __init__(self, operation_mode=OperationMode.both, name=None):
+        """
+            Constructor. Initializes graph variables.
+
+            Args:
+                operation_mode: Graph operation mode, that will be propagated along modules during graph creation.
+                [training | eval | both] (DEFAULT: both)
+                name: Name of the graph (optional)
+        """
+        # Initialize the inferface.
+        super().__init__(name)
+
+        # Register graph.
+        self._name = self._app_state.register_graph(self, name)
+
+        # Store name and operation mode.
+        self._operation_mode = operation_mode
+
+        # "Modules" - list of modules constituting "nodes" in a given graph.
+        self._modules = {}
+
+        # All tensors produced within this graph (dict of dicts).
+        # This stores  "all output tensors" dictionary, where the key is the name of "producer" module,
+        # and the value contains a dictionary of all tensors produced by it.
+        self._all_tensors = {}
+
+        # "Steps": order of the  execution of modules in a graph.
+        self._steps = OrderedDict()
+
+        # Bound inputs.
+        self._inputs = GraphInputs()
+
+        # Bound outputs.
+        self._outputs = GraphOutputs(self._all_tensors)
+
```

Comment:
Why are these bound with NeuralGraph is initialized? Wouldn't an initialized NeuralGraph always be empty?

Reply:
Right. And initially a graph instance has empty lists of bound inputs and outputs...

hmmm, not sure if I understand the question here...?